### DISCUSSION TITLE
unique key constraint validation

### DISCUSSION THREAD

**Discussion by idanfishman** [fastapi/fastapi#4980] [Questions] [answered]
Hi Guys,
Coming from Django and DRF unique key validation is a piece of cake.
I'm trying to understand what is the best practice to this kind of validation.

I have some keys that need to be unique within the db table, but I can't figure out how to do so.
I looked for answer in pydantic's repo and they are say it is a bad practice to have a validator on the schema which queries the database and checks for duplicate.

so, what is the best practice to validate a unique key value? 

example case:
The values of the keys `name` and `business_number` should be unique.

```python
    # crud create
    def create(self, db: Session, obj: CreateSchemaType) -> ModelType:
        obj_data = jsonable_encoder(obj)
        db_obj = self.model(**obj_data)
        db.add(db_obj)
        db.commit()
        db.refresh(db_obj)
        return db_obj
    
    # model
    class Company(Base):
        __tablename__ = "companies"
    
        id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    
        name = Column(String, nullable=False, index=True)
        business_number = Column(String, nullable=False, index=True)
        carrier_license_exp_date = Column(Date)
        professional_manager = Column(JSON, nullable=False)
    
        address = Column(JSON)
        contact = Column(JSON)
        email = Column(String)
        fax_number = Column(String)
        phone_number = Column(String)

    # schema
    class CompanyBase(BaseModel):
        name: str
        business_number: str
        professional_manager: ProfessionalManager
    
        carrier_license_exp_date: date | None
        contact: Contact | None = None
        address: Address | None = None
        phone_number: str | None = None
        fax_number: str | None = None
        email: EmailStr | None = None

    # endpoint
    @router.post(
        "/",
        status_code=status.HTTP_201_CREATED,
        response_model=schemas.Company,
        response_model_exclude_none=True,
    )
    async def create_company(
        company_obj: schemas.CompanyCreate,
        db: Session = Depends(deps.get_db),
    ) -> Any:
        company = crud.company.create(db=db, obj=company_obj)
        return company
```

**Comment by odiseo0** [comment, marked as answer]
I think the right way is to go to your database and create a `Unique Constraint` in both columns, I think it's the simplest and best way to do so

**Reply by idanfishman** [reply]
Creating the unique constraint on db level, causes internal server error 500. unless you try and except this error which is a bad practice.
 
Before inserting data to db you should.
1. Value Validation -> pydantic
2. DB integrity Validation -> ? 
3. Insert to DB -> sqlalchemy

**Reply by JarroVGIT** [reply]
Why is catching an exception bad practice? If you make it a specific exception that you want to catch, it is completely okay to do so. 

It is the whole reason DBs offer this functionality, I would say.
